featurebyte/api/user_defined_function.py

```diff
@@ -0,0 +1,319 @@
+                continue
+            name_to_udf_dict[udf.name] = udf
+    except InvalidSettingsError:
+        # ignore invalid settings error due to fail to connect to the server
+        logger.info("Failed to synchronize user-defined functions.")
+
+    # construct & add all user-defined functions to the function accessor
+    for udf in name_to_udf_dict.values():
+        UserDefinedFunctionInjector.create_and_add_function(func_accessor, udf)
+
+
+class FunctionDescriptor:
+    """
+    FunctionDescriptor class used to perform dynamic function registration when the descriptor is accessed.
+    """
+
+    def __init__(self, route: str):
+        self._route = route
+
+    def __get__(self, instance: object, owner: type[object]) -> FunctionAccessor:
+        # create a function accessor object as a user-defined function container and
+        # register all user-defined functions to the function accessor
+        accessor = FunctionAccessor()
+        _synchronize_user_defined_function(
+            func_accessor=accessor, route=self._route, feature_store_id=None
+        )
+        return accessor
+
+
+class UserDefinedFunction(DeletableApiObject, SavableApiObject):
+    """
+    UserDefinedFunction class used to represent a UserDefinedFunction in FeatureByte.
+    """
+
+    # class variables
+    _route = "/user_defined_function"
+    _create_schema_class = UserDefinedFunctionCreate
+    _get_schema = UserDefinedFunctionModel
+    _list_schema = UserDefinedFunctionModel
+    _list_fields = [
+        "signature",
+        "function_name",
+        "feature_store_name",
+        "is_global",
+    ]
+    _list_foreign_keys = [
+        ForeignKeyMapping("feature_store_id", FeatureStore, "feature_store_name"),
+    ]
+
+    # pydantic instance variable (internal use)
+    internal_catalog_id: Optional[PydanticObjectId] = Field(alias="catalog_id")
+    internal_feature_store_id: PydanticObjectId = Field(alias="feature_store_id")
+    internal_function_name: str = Field(alias="function_name")
+    internal_function_parameters: List[FunctionParameter] = Field(alias="function_parameters")
+    internal_output_dtype: DBVarType = Field(alias="output_dtype")
+
+    # function accessor containing all user-defined functions
+    func: ClassVar[Any] = FunctionDescriptor(route=_route)
+
+    @property
+    def catalog_id(self) -> Optional[PydanticObjectId]:
+        """
+        Catalog id of the user-defined function. If the user-defined function is global, the catalog id is None.
+
+        Returns
+        -------
+        Optional[PydanticObjectId]
+        """
+        return self.cached_model.catalog_id
+
+    @property
+    def feature_store_id(self) -> PydanticObjectId:
+        """
+        Associated feature store id of the user-defined function.
+
+        Returns
+        -------
+        PydanticObjectId
+        """
+        return self.cached_model.feature_store_id
+
+    @property
+    def function_name(self) -> str:
```

Comment:
Would it be clearer to name this `sql_function_name` to distinguish this from the user defined function's `name`?

Reply:
Create a ticket to update this (https://featurebyte.atlassian.net/browse/DEV-1880) as it would be better to fix it at the model level too. Is it ok to you?

Reply:
Sure 👍